Integrate tour-kit library, implement Course and Lesson tours

Imran92 opened the pull request:
Resolves https://github.com/Automattic/sensei/issues/7444
And https://github.com/Automattic/sensei/issues/7477
Part of https://github.com/Automattic/sensei/issues/7494


## Proposed Changes
* Integrated [tour-kit](https://www.npmjs.com/package/@automattic/tour-kit) library to show sensei tours
* Added Tours for Courses and Lessons

### Separate Issues:
I'll create separate mini issues to deal with the following ones (and anything else we notice while testing this) to not deal with them here in this PR and enable working on other issues that are dependant on the tour kit. As this feature is behind a feature flag, we can go ahead and merge it if it looks good overall, and can work the other issues separately. The main goal of this PR is integrating the tour-kit library properly.

- We don't deal with the loading logic here. We are loading Course and Lesson tours for all Courses and Lessons
- When a tour is closed by clicking on the cross, we don't persist that data for that user, so it'll load again on next reload
- Show tour only after modal is closed (Update - later I implemented this and also introduced a redux store, to fix an e2e issue. it still needs some work and I'll also add tests under the dedicated issue for this https://github.com/Automattic/sensei/issues/7494)
- When content drip is enabled, tour is shown over the content drip area below.

## Testing Instructions
<!--
Add as many details as possible to help others reproduce the issue and test the changes.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

1. Make sure your local environment is set up as 'development' [env](https://developer.wordpress.org/reference/functions/wp_get_environment_type/). Otherwise, this feature won't be enabled
2. Run `composer install`
3. Run `npm install`
4. Make sure you're running "npm run start" or have run "npm run build"
5. Create a new Course and make sure you can take the tour after adding a Course Outline block, minimize it and cross it
6. Create a new lesson and make sure you see the tour after adding a Quiz block, minimize it and cross it

#### Course Tour
https://github.com/Automattic/sensei/assets/6820724/825eef9e-45b0-4f85-85bf-74f8c08e82fe

#### Lesson Tour
https://github.com/Automattic/sensei/assets/6820724/b152b239-dc8b-4962-a767-0ed4d1411f03



## New/Updated Hooks
<!-- Add the following only if there are new/updated actions or filters. Please provide a brief description of what they do and any arguments they may take. Be sure to also add the "Hooks" label to this PR. -->

* `sensei_tour_loaders` - Loads a sensei tour script

## Pre-Merge Checklist
<!-- Complete applicable items on this checklist **before** merging. Items that are not applicable can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed. -->
- [ ] PR title and description contain sufficient detail and accurately describe the changes
- [x] Acceptance criteria is met
- [ ] Decisions are publicly documented
- [x] Adheres to coding standards ([PHP](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/), [JavaScript](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/), [CSS](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/css/), [HTML](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/html/))
- [x] All strings are translatable (without concatenation, handles plurals)
- [x] Follows our naming conventions (P6rkRX-4oA-p2)
- [x] Hooks (p6rkRX-1uS-p2) and functions are documented
- [ ] New UIs are responsive and use a [mobile-first approach](https://zellwk.com/blog/how-to-write-mobile-first-css/)
- [x] New UIs match the designs
- [ ] Different user privileges (admin, teacher, subscriber) are tested as appropriate
- [ ] Legacy courses (course without blocks) are tested
- [x] Code is tested on the minimum supported PHP and WordPress versions
- [ ] User interface changes have been tested on the latest versions of Chrome, Firefox and Safari
- [ ] "Needs Documentation" label is added if this change requires updates to documentation
- [ ] Known issues are created as new GitHub issues

Imran92 committed:
Add libraries for tourkit build

Imran92 committed:
Add empty babel config file to fix webpack file error

Imran92 committed:
load tour-kit library

Imran92 committed:
Update versions to match needs of tour kit

Imran92 committed:
Update package lock file

Imran92 committed:
Add config to remove parsing errors

Imran92 committed:
load files in webpack

Imran92 committed:
Add sensei tourkit component

Imran92 committed:
Move course tour steps

Imran92 committed:
Add course tour component

Imran92 committed:
Add lesson steps

Imran92 committed:
Add lesson tour component

Imran92 committed:
Add styles for tour

Imran92 committed:
Add type

Imran92 committed:
Add sensei tour loader class

Imran92 committed:
Load tour class

Imran92 committed:
Update type

Imran92 committed:
Use namespace and remove type

Imran92 committed:
Load scripts only in correct page hooks

Imran92 committed:
Update hook doc, load style only with tour

Imran92 committed:
Import namespace in loader class

Imran92 committed:
Test loader hook getting attached on init

Imran92 committed:
test scripts don't get loaded for other post types

Imran92 committed:
Test course scripts get loaded properly

Imran92 committed:
Test lesson scripts get loaded for lesson CPT

Imran92 committed:
Test lesson scripts get loaded for existing lesson edit

Imran92 committed:
Scripts should not get loaded for other pages

Imran92 committed:
Test hook modification

Imran92 committed:
Fix tests

Imran92 committed:
Remove jest config

Imran92 committed:
Update library

Imran92 committed:
Update packages

Imran92 committed:
Update classname package to prevent double version

Imran92 committed:
Test that course tour component loads when needed

Imran92 committed:
Check that step params are properly passed

Imran92 committed:
Test lesson tour renderer

Imran92 committed:
Should render lesson tour when quiz block is loaded

Imran92 committed:
Fix php lint

Imran92 committed:
Fix psalm issue

Imran92 committed:
Merge branch 'trunk' into add/tour-kit-lib

Imran92 committed:
Add test to make sure tour kit renders

Imran92 committed:
Test that proper steps are passed

Imran92 committed:
Test that addition config is properly merged

Imran92 committed:
Load styles from components library instead

Imran92 committed:
Add next version and internal tags

Imran92 committed:
Merge branch 'trunk' into add/tour-kit-lib

Imran92 commented:
E2E issue didn't appear before because of CI failure. Looking at it now

Imran92 committed:
Add tour redux store

Imran92 committed:
Load redux store, show tour conditionally

Imran92 committed:
Don't show course tour when modal is open

Imran92 committed:
Merge branch 'add/tour-kit-lib' of https://github.com/Automattic/sensei into add/tour-kit-lib

Imran92 committed:
Merge branch 'trunk' into add/tour-kit-lib

Imran92 committed:
Update e2e test

Imran92 committed:
Merge branch 'trunk' into add/tour-kit-lib

Imran92 commented:
> E2E issue didn't appear before because of CI failure. Looking at it now

All fixed now

Imran92 committed:
Merge branch 'trunk' into add/tour-kit-lib

Imran92 committed:
Merge branch 'trunk' into add/tour-kit-lib

Imran92 committed:
Add only local argument like calypso

Imran92 committed:
Use only local argument when closing tour locally

Imran92 committed:
Load only incomplete tours

Imran92 committed:
Add function to set tour status

Imran92 committed:
Add function to get tour status

Imran92 committed:
Add controller for setting tour status for user

Imran92 committed:
Load tour api

Imran92 committed:
Update component to take in unique tour id

Imran92 committed:
Update course and lesson tour to pass tour id

Imran92 committed:
Add test for setting meta properly

Imran92 committed:
Add test for existing meta update for tour

donnapep commented:
Gave this a quick test and it's looking good! One thing that is a bit off is the quotation marks in some of the text. I'm guessing you copied them right from the Google doc? If so, I'm guessing that's the issue.

> I'll create separate mini issues to deal with the following ones

Do we have an issue for ensuring that the height stays the same? Right now the box is jumping around. Maybe we could use a height that accommodates 3 lines and I'll shorten any lines that are more than 3?

I couldn't get the lesson tour to work for a new or existing lesson, only the course tour. How can I get it to work?

Imran92 committed:
Merge branch 'trunk' into add/tour-kit-lib

Imran92 committed:
Merge branch 'add/tour-kit-lib' into add/persist-tour-show-choice

Imran92 committed:
Add store test for data setting

Imran92 committed:
Make sure API fetch is called when show is false

Imran92 committed:
Should not call api for local

Imran92 committed:
Should set complete = false when show = true

Imran92 committed:
Check if value locally sets as expected when param not passed

Imran92 committed:
Test that close button is called as expected

Imran92 committed:
Test that meta setting function works as expected

Imran92 committed:
Add test for updating existing meta

Imran92 committed:
Add test for getter function for tour meta

Imran92 committed:
Fix psalm errors

Imran92 committed:
Add fixed height for tour modal

Imran92 committed:
Change quotation marks

Imran92 commented:
Thanks so much for checking it @donnapep !

> Do we have an issue for ensuring that the height stays the same? Right now the box is jumping around. Maybe we could use a height that accommodates 3 lines and I'll shorten any lines that are more than 3?

I added the styles in this PR initially but later it got removed unintentionally during some refactoring. Thanks for noticing! I've set a fixed height here, this is the height it sets for itself when it has 3 lines in it, so I've forced it here https://github.com/Automattic/sensei/pull/7487/commits/135cf82a71e089e13fdd641a772986f00d910c31 using CSS (I've also removed a bit of the padding from top, Wpcomtourkit added it probably because it looks good when they have images, for us the space looks a bit too much, LMK if you think otherwise). I've noticed that all of our contents are withing 3 lines. Except for only one in the quiz tour - 
<img width="400" alt="Screenshot 2024-02-27 at 7 10 06 PM" src="https://github.com/Automattic/sensei/assets/6820724/162fdee7-d693-4bd7-80f7-90b34c323949">

> I'm guessing you copied them right from the Google doc? If so, I'm guessing that's the issue.

You're right, I've updated the quotation marks here. Thanks! https://github.com/Automattic/sensei/pull/7487/commits/49fe2077b4641a0007bf032cd916be63d022da80

> I couldn't get the lesson tour to work for a new or existing lesson, only the course tour. How can I get it to work?

Oh, it requires having a Quiz block. I showed it in the attached video above before, now I've also updated the instructions mentioning it.

Thanks!

Imran92 committed:
Merge branch 'add/tour-kit-lib' into add/persist-tour-show-choice

donnapep commented:
> Oh, it requires having a Quiz block.

Oh, of course! 🤦🏻‍♀️ I see it now.

> I've noticed that all of our contents are withing 3 lines. Except for only one in the quiz tour

I checked and I don't think I'm going to shorten that one given that it looks reasonable and the text isn't being cut off.

I'm not approving this yet as I haven't done a code review, but everything looks great from a UI / functionality perspective. Nice work! ❤️

Imran92 committed:
Merge branch 'trunk' into add/tour-kit-lib

Imran92 committed:
Load a tour only when install version is higher than required

Imran92 committed:
Test that tour is loaded properly according to version

renatho commented on assets/admin/tour/lesson-tour/index.js:
```diff
@@ -0,0 +1,30 @@
+export const getQuizBlock = () =>
+	getFirstBlockByName(
+		'sensei-lms/quiz',
+		select( 'core/block-editor' ).getBlocks()
```
I wonder if we could use the `useSelect` hook here instead to be reactive. So it can avoid some issues in the future in case it doesn't re-render. Also, I noticed the tour is displayed only when losing the focus on the quiz block. I suspect it would be fixed with this change too.

renatho replied:
And the same would apply to courses.

Imran92 replied:
Good point! I've updated it here https://github.com/Automattic/sensei/pull/7487/commits/41c45dc96abb699012304bd401df0ce77187e2c3 and here https://github.com/Automattic/sensei/pull/7487/commits/2615108a5d699fc78b5744cc2ea655126cc77d05 using `useSelect`. Thanks!

renatho commented on assets/admin/editor-wizard/editor-wizard-modal.js:
```diff
@@ -61,6 +70,7 @@ const EditorWizardModal = () => {
 	}, [] );
 
 	useEffect( () => {
+		showSenseiTour( false );
```
Could we avoid setting to `false` and later set to `true` if we set the default value to `false` in the store?

Imran92 replied:
Yap we can do that, but in our case, we're assuming (till now) that if a tour is loaded, it's meant to be shown, unless some component wants it otherwise. For example, we show this Tour even when the modal is not shown (in the scenario, when an existing Course is opened for editing, not creating a new course). So as the editor wizard modal will not be there in that case, so we'll need to set it to True from some other place. So that's why I've kept it like it is now. But these are not very strong reasons and I don't have a strong preference. We can change it later too if different use cases come up.

renatho replied:
Okay! ;)

renatho commented on assets/admin/editor-wizard/editor-wizard-modal.js:
```diff
@@ -34,12 +36,19 @@ const EditorWizardModal = () => {
 	} );
 
+	const showSenseiTour = ( show ) => {
+		if ( setTourShowStatus ) {
```
Is this check and the `?? {}` really needed? Maybe some case I'm missing here?

Imran92 replied:
Good point, I've removed it here https://github.com/Automattic/sensei/pull/7487/commits/118b4742831533066e384affa9545c11b195173e. Thanks!

renatho replied:
Just one more thing related to this. I think `if ( setTourShowStatus ) {` is probably not needed too, right?

Imran92 replied:
Thanks @renatho ! I've removed it here https://github.com/Automattic/sensei/pull/7487/commits/5e4fb9b60419bf59e0edf4ecec61722731c24b6c

renatho commented on assets/admin/tour/components/sensei-tour-kit/index.js:
```diff
@@ -0,0 +1,69 @@
+/**
+ * Renders a tour kit component using Sensei.
+ *
+ * @param {Object}     props                  - Component props.
```
There are some extra `-` in the JSDoc here. :)

Imran92 replied:
Thanks! I've removed it here https://github.com/Automattic/sensei/pull/7487/commits/bb3862fcb5aa5c927257bd9358ee1f107ea967c5

renatho commented on assets/admin/tour/components/sensei-tour-kit/index.js:
```diff
@@ -0,0 +1,69 @@
+	const { showTour } = useSelect( ( select ) => {
+		const { getIfShowTour } = select( SENSEI_TOUR_STORE );
+		return {
+			showTour: getIfShowTour(),
```
Just a suggestion. While reading the code, I got a little confused about the `getIfShowTour` selector name. I'd suggest something like `getShowTour` or `shouldShowTour`. WDYT?

Imran92 replied:
Good suggestion! I've updated it here https://github.com/Automattic/sensei/pull/7487/commits/3c783ba04b4f53c40d8cbfc5e499528dc80f2929

renatho reviewed (commented):
I started reviewing in the wrong order, haha. So I think some of the comments I addded in https://github.com/Automattic/sensei/pull/7510 would fit more here.

renatho commented:
@Imran92, for the reviews I'm doing in your chain of PRs, if you want to merge earlier and work on the items later, I think that's fine. Just suggesting this because I know how hard it is to change the PRs in a chain. haha

Imran92 commented:
> @Imran92, for the reviews I'm doing in your chain of PRs, if you want to merge earlier and work on the items later, I think that's fine. Just suggesting this because I know how hard it is to change the PRs in a chain. haha

Hahah! Yes, that'll be great, it's indeed a big headache to keep maintaining chained PRs. I'm merging the other two in this one. I'll address them in this PR. Thanks!!

Imran92 committed:
Merge pull request #7519 from Automattic/add/show-tour-only-for-new-instance

Show tours only for new Sensei installations

Imran92 committed:
Merge pull request #7510 from Automattic/add/persist-tour-show-choice

When tour is closed or skipped, persist the choice and don't load that tour for that user

Imran92 committed:
Merge branch 'trunk' into add/tour-kit-lib

Imran92 committed:
Fix after merge

Imran92 committed:
Exhaust dependency

Imran92 committed:
Add tracking centrally in tour component

Imran92 committed:
Enable tour for Course tour

Imran92 committed:
Enable track for lesson tour

Imran92 committed:
Add test that track event is properly called by callback

Also should contain the step data and tracks id

Imran92 committed:
Test that event log is not called without track id

Imran92 committed:
test course is passing tracking id

Imran92 committed:
test that lesson tour is passing track id

Imran92 committed:
Use useSelect in Course tour

Imran92 committed:
Use useSelect in Course tour

Imran92 committed:
Use useSelect in Lesson tour

Imran92 committed:
Use useSelect in Lesson tour

Imran92 committed:
Fix course tour tests

Imran92 committed:
Fix lesson tour test

Imran92 committed:
Fix jsDoc extra dash

Imran92 committed:
Remove extra check

Imran92 committed:
Remove extra check

Imran92 committed:
rename getIfShowTour to shouldShowTour

Imran92 committed:
rename getIfShowTour to shouldShowTour

Commit:
Fix course tour tests

Commit:
Fix lesson tour test

Commit:
Fix jsDoc extra dash

Imran92 committed:
Merge branch 'add/tour-kit-lib' into add/track-course-lesson-tour-steps

renatho reviewed (dismissed):
Thank you for the tweaks and explanations, @Imran92!

It looks good and works well to me! I added just a final comment iterating in a comment, but I'm already approving if you want to change that and merge it immediately. 😉

Imran92 committed:
Update step name to be welcome

Imran92 committed:
Remove callback function

Imran92 committed:
Merge branch 'add/tour-kit-lib' into add/track-course-lesson-tour-steps

Imran92 committed:
Fix multiple call to logging

Imran92 committed:
Merge pull request #7522 from Automattic/add/track-course-lesson-tour-steps

Add tracks event in Sensei tour component and track Course and Lesson onboarding tour steps
